### PR TITLE
feat: onBarCodeRead should also send raw photo bytes, fixed #2786

### LINKED ISSUE #2786: onBarCodeRead should also send raw photo bytes
[Issuehunt badges](https://issuehunt.io/r/react-native-community/react-native-camera/issues/2786)

# Feature-Request

**Describe the Feature**

Currently `onBarCodeRead` only gives metadata, but not the image (bytes/base64) itself.
Calling `takePictureAsync()` does not give the real photo that was detected.

**Possible Implementations**

1. Save the image(s) in a bounded queue, make an optional prop to set format, i.e. `detectedImageInEvent` = `false` (default) | `true` (will be encoded as base64). (I thought of adding `binary` option but seems like [base64 is a better idea](https://stackoverflow.com/a/1443240/122441))
   * `false` is default as it adds significant additional processing
2. If barcode is detected, provide the `image` to the callback handler.

**Related Issues**

#2763, #2018.

_Bonus points:_ Support also `onGoogleVisionBarcodesDetected`, but this turns out to be a different processing pipeline.

> Love react-native-camera? Please consider supporting our collective: 👉  https://opencollective.com/react-native-camera/donate
> Want this feature to be resolved faster? Please consider adding a bounty to it https://issuehunt.io/repos/33218414

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $10.00)

- [ceefour ceefour](https://issuehunt.io/u/ceefour) ($10.00)

### Submitted pull Requests
- [#2923 feat: onBarCodeRead should also send raw photo bytes, fixed #2786](https://issuehunt.io/r/react-native-community/react-native-camera/pull/2923)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/react-native-community/react-native-camera/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by ceefour**
# Summary

onBarCodeRead sends raw photo bytes, fixed issue #2786.

## Test Plan

See discussion in #2786.

### What's required for testing (prerequisites)?

See discussion in #2786.

### What are the steps to reproduce (after prerequisites)?

See discussion in #2786.

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ❌     |
| Android |    ✅     |

## Checklist

- [x] I have tested this on a device and a simulator
- [x] I added the documentation in `docs/RNCamera.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [x] I updated the typed files (TS and Flow)
- [ ] I added a sample use of the API in the example project (`example/App.js`)

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#2786: onBarCodeRead should also send raw photo bytes](https://issuehunt.io/repos/33218414/issues/2786)
---
</details>

### CODE REVIEW

**Comment by sibelius** [comment]
how this affects performance?

**Comment by ceefour** [comment]
@sibelius It doesn't, it's optional and need to be explicitly enabled as described in #2786.

File: android/src/main/java/org/reactnative/camera/RNCameraView.java [hunk 1]
```diff
@@ -508,11 +534,28 @@ public void setGoogleVisionBarcodeMode(int barcodeMode) {
...
 
-  public void onBarcodesDetected(WritableArray barcodesDetected) {
+  public void onBarcodesDetected(WritableArray barcodesDetected, int width, int height, byte[] imageData) {
     if (!mShouldGoogleDetectBarcodes) {
       return;
     }
-    RNCameraViewHelper.emitBarcodesDetectedEvent(this, barcodesDetected);
+
+    // See discussion in https://github.com/react-native-community/react-native-camera/issues/2786
+    final byte[] compressedImage;
```
Comment (sibelius) [inline android/src/main/java/org/reactnative/camera/RNCameraView.java:543]: is this logic the same as above?

should we extract this to a method/function?

Comment (ceefour) [reply android/src/main/java/org/reactnative/camera/RNCameraView.java:543]: Yes, similar logic but different barcode reader providers in Android.

**Comment by ceefour** [comment]
@sibelius I pushed a commit with documentation in `docs/RNCamera.md`

**Pull request merged**